hibernate-core/src/main/java/org/hibernate/cfg/annotations/PropertyBinder.java

```diff
@@ -190,6 +197,31 @@ private Property makePropertyAndValue() {
 		);
 		basicValueBinder.setReferencedEntityName( referencedEntityName );
 		basicValueBinder.setAccessType( accessType );
+
+
+		if ( property.isAnnotationPresent(TenantId.class) ) {
+			InFlightMetadataCollector collector = buildingContext.getMetadataCollector();
+			collector.addFilterDefinition(
+					new FilterDefinition(
+							"_tenantId",
+							"",
+							singletonMap(
+									"tenantId",
+									collector.getTypeConfiguration()
+											.getBasicTypeRegistry()
+											.getRegisteredType(String.class)
```

Comment:
It would be cool if we could support other kinds of identifiers as well i.e. integers are very common.

Reply:
Unfortunately, we would also need to change a few other APIs that currently expect a String, but I think this should be a separate issues and PR.

Reply:
FTR, moved to https://hibernate.atlassian.net/browse/HHH-14822